File: golang/leetcode/typescript/search_range.ts
const searchRange = (nums: number[], target: number): number[] => {
    const start = findStart(nums, target);
    const end = findEnd(nums, target);
    return [start, end];
    };

const findStart = (nums: number[], target: number): number => {
    let start = 0;
    let end = nums.length - 1;
    while (start <= end) {
        const mid = Math.floor((start + end) / 2);
        if (nums[mid] < target) {
            start = mid + 1;
        } else if (nums[mid] > target) {
            end = mid - 1;
        } else {
            if (mid === 0 || nums[mid - 1] !== target) {
                return mid;
            }
            end = mid - 1;
        }
    }
    return -1;
};

const findEnd = (nums: number[], target: number): number => {
    let start = 0;
    let end = nums.length - 1;
    while (start <= end) {
        const mid = Math.floor((start + end) / 2);
        if (nums[mid] < target) {
            start = mid + 1;
        } else if (nums[mid] > target) {
            end = mid - 1;
        } else {
            if (mid === nums.length - 1 || nums[mid + 1] !== target) {
                return mid;
            }
            start = mid + 1;
        }
    }
    return -1;
};

// test
const nums = [5, 7, 7, 8, 8, 10];
const target = 8;
console.log(searchRange(nums, target));

File: golang/leetcode/rust/search_range.rs
fn searchRange( nums: Vec<i32>, target: i32) -> Vec<i32> {
    let mut l = 0;
    let mut r = nums.len() - 1;
    let mut m = 0;
    while l <= r {
        m = (l + r) / 2;
        if nums[m] == target {
            return vec![m as i32, m as i32];
        }
        if nums[l] <= nums[m] {
            if nums[l] <= target && target < nums[m] {
                r = m - 1;
            } else {
                l = m + 1;
            }
        } else {
            if nums[m] < target && target <= nums[r] {
                l = m + 1;
            } else {
                r = m - 1;
            }
        }
    }
    vec![-1, -1]
}